lib/generic.ts

```diff
@@ -28,6 +30,80 @@ type GenericProvable<T, TValue, Field> = {
   toValue: (x: T) => TValue;
   fromValue: (x: T | TValue) => T;
 };
+
+type StructProvable<T extends {}, TValue extends {[K in keyof T]: any}, Field> = {
+    [K in keyof T]: GenericProvable<T[K], TValue[K], Field>;
+}
+
+function structProvable<T extends {}, TValue extends {[K in keyof T]: any}, Field>
+  (typ: StructProvable<T, TValue, Field>):
+      GenericProvable<
+        {[K in keyof T]: T[K]},
+        {[K in keyof TValue]: TValue[K]},
+        Field>
+{
+    return {
+        toFields: (x) => {
+            var fields: Field[] = [];
+            var field: keyof T;
+            for (field in typ) {
+                fields.concat(typ[field].toFields(x[field]));
+            }
+            return fields;
+        },
+        toAuxiliary: (x? /* Why is this nullable?! */) => {
```

Comment:
it's supposed to be called without argument, and return a default value, when we're in compilation mode